File: programmers/lv0/코딩기초트레이닝/문자열겹쳐쓰기/Solution.cpp
#include <string>
#include <vector>

using namespace std;

// mine
string solution(string my_string, string overwrite_string, int s) {
    vector<char> v(my_string.length());
    for (int i = 0; i < my_string.length(); ++i) {
        v[i] = my_string[i];
    }
    for (int i = 0, j = s; i < overwrite_string.length(); ++i, ++j) {
        v[j] = overwrite_string[i];
    }
    string answer(v.begin(), v.end());
    return answer;
}

// simple
string solution(string my_string, string overwrite_string, int s) {
    return my_string.replace(s,overwrite_string.size(),overwrite_string);
}

File: programmers/lv0/코딩기초트레이닝/문자열겹쳐쓰기/Solution.swift
import Foundation

func solution(_ my_string:String, _ overwrite_string:String, _ s:Int) -> String {
    // 1번째 방법
    let result = String(my_string.prefix(s)) + overwrite_string + String(my_string.suffix(my_string.count - (s+overwrite_string.count)))

    // 2번째 방법
    var answer = my_string
    let startIndex = answer.index(answer.startIndex, offsetBy: s)
    let endIndex = answer.index(startIndex, offsetBy: overwrite_string.count)
    let range = startIndex..<endIndex
    answer.replaceSubrange(range, with: overwrite_string)

    // String Range 숫자 디버깅방법
//    let distance1 = answer.distance(from: answer.startIndex, to: startIndex)
//    let distance2 = answer.distance(from: answer.startIndex, to: endIndex)
//    let distance3 = answer.distance(from: startIndex, to: endIndex)
//    print(distance1)
//    print(distance2)
//    print(distance3)


    return answer
}

print(solution("He11oWor1d", "lloWorl", 2))